mathics/eval/datetime.py

```diff
@@ -0,0 +1,94 @@
+"""
+Implementation of the evaluation of datetime related expressions.
+"""
+
+import sys
+import time
+from datetime import datetime
+from typing import Optional
+
+from mathics.core.element import BaseElement
+from mathics.core.evaluation import Evaluation
+
+PAUSE_TICKS_PER_SECOND = 1000
+
+
+def valid_time_from_expression(t: BaseElement, evaluation: Evaluation) -> float:
+    """
+    Try to evaluate t to a nonnegative float number.
+    """
+    t = t.evaluate(evaluation)
+    if not t.is_numeric(evaluation):
+        raise ValueError
+    try:
+        timeout = float(t.to_python())
+    except TypeError:
+        raise ValueError
+    if timeout < 0:
+        raise ValueError
+    return timeout
+
+
+def eval_pause(sleeptime: float, evaluation):
+    """
+    Do a pause for `sleeptime`. If sleep
+    """
+    # Due to the GIL lock, if we implement this method
+    # by just calling `time.sleep(sleeptime)`, the
+    # evaluation would not be aware of `TimeoutException`
+    # raised by an outer `TimeConstrained` expression.
+    #
+    # For this reason, the splitting of the sleep time is
+    # needed.
+    # It was also noticed in tests that in some platforms
+    # the total time that takes n calls to time.sleep(delta_t)
+    # can be appreciably larger than n*delta_t. For this reason,
+    # we also need to check that inside the loop that the
+    # enlapsed time at the i-esim iteration does not exceed
+    # the  desired total time.
+
```

Comment:
Perhaps mention the possibility of implementing using `Threading.Event()`

Reply:
I put the comment in mathics.eval.datetime.